File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/GodUpdateRegionInfo.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.UUID;

/**
 * GodUpdateRegionInfo
 * Sent from viewer to sim after a god has changed some
 * of the parameters in the god tools floater
 * viewer -> sim
 * reliable
 *
 * <p>Template: {@code GodUpdateRegionInfo Low 143 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
public class GodUpdateRegionInfo extends SLMessage {
    public AgentData AgentData_Field;
    public ArrayList<RegionInfo2> RegionInfo2_Fields = new ArrayList<>();
    public RegionInfo RegionInfo_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public UUID SessionID; // LLUUID
    }

    /** Block RegionInfo, Single. */
    public static class RegionInfo {
        public float BillableFactor; // F32
        public int EstateID; // U32
        public int ParentEstateID; // U32
        public int PricePerMeter; // S32
        public int RedirectGridX; // S32
        public int RedirectGridY; // S32
        public int RegionFlags; // U32
        public byte[] SimName; // Variable 1 - string
    }

    /** Block RegionInfo2, Variable. */
    public static class RegionInfo2 {
        public long RegionFlagsExtended; // U64
    }

    public GodUpdateRegionInfo() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
        this.RegionInfo_Field = new RegionInfo();
    }

    @Override
    public int CalcPayloadSize() {
        return this.RegionInfo_Field.SimName.length + 1 + 4 + 4 + 4 + 4 + 4 + 4 + 4 + 36 + 1 + (this.RegionInfo2_Fields.size() * 8);
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleGodUpdateRegionInfo(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 143 (GodUpdateRegionInfo).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x8F);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packVariable(byteBuffer, this.RegionInfo_Field.SimName, 1);
        packInt(byteBuffer, this.RegionInfo_Field.EstateID);
        packInt(byteBuffer, this.RegionInfo_Field.ParentEstateID);
        packInt(byteBuffer, this.RegionInfo_Field.RegionFlags);
        packFloat(byteBuffer, this.RegionInfo_Field.BillableFactor);
        packInt(byteBuffer, this.RegionInfo_Field.PricePerMeter);
        packInt(byteBuffer, this.RegionInfo_Field.RedirectGridX);
        packInt(byteBuffer, this.RegionInfo_Field.RedirectGridY);
        byteBuffer.put((byte) this.RegionInfo2_Fields.size());
        Iterator<?> it = this.RegionInfo2_Fields.iterator();
        while (it.hasNext()) {
            packLong(byteBuffer, ((RegionInfo2) it.next()).RegionFlagsExtended);
        }
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.RegionInfo_Field.SimName = unpackVariable(byteBuffer, 1);
        this.RegionInfo_Field.EstateID = unpackInt(byteBuffer);
        this.RegionInfo_Field.ParentEstateID = unpackInt(byteBuffer);
        this.RegionInfo_Field.RegionFlags = unpackInt(byteBuffer);
        this.RegionInfo_Field.BillableFactor = unpackFloat(byteBuffer);
        this.RegionInfo_Field.PricePerMeter = unpackInt(byteBuffer);
        this.RegionInfo_Field.RedirectGridX = unpackInt(byteBuffer);
        this.RegionInfo_Field.RedirectGridY = unpackInt(byteBuffer);
        int i = byteBuffer.get() & 0xFF;
        for (int i2 = 0; i2 < i; i2++) {
            RegionInfo2 regionInfo2 = new RegionInfo2();
            regionInfo2.RegionFlagsExtended = unpackLong(byteBuffer);
            this.RegionInfo2_Fields.add(regionInfo2);
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/GodUpdateRegionInfo.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.ArrayList
import java.util.UUID

/**
 * GodUpdateRegionInfo
 * Sent from viewer to sim after a god has changed some
 * of the parameters in the god tools floater
 * viewer -> sim
 * reliable
 *
 * <p>Template: {@code GodUpdateRegionInfo Low 143 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
open class GodUpdateRegionInfo : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField val RegionInfo2_Fields = ArrayList<RegionInfo2>()
    @JvmField var RegionInfo_Field: RegionInfo = RegionInfo()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var SessionID: UUID? = null
    }

    /** Block RegionInfo, Single. */
    open class RegionInfo {
        @JvmField var BillableFactor: Float = 0f
        @JvmField var EstateID: Int = 0
        @JvmField var ParentEstateID: Int = 0
        @JvmField var PricePerMeter: Int = 0
        @JvmField var RedirectGridX: Int = 0
        @JvmField var RedirectGridY: Int = 0
        @JvmField var RegionFlags: Int = 0
        @JvmField var SimName: ByteArray? = null
    }

    /** Block RegionInfo2, Variable. */
    open class RegionInfo2 {
        @JvmField var RegionFlagsExtended: Long = 0L
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return RegionInfo_Field.SimName!!.size + 1 + 4 + 4 + 4 + 4 + 4 + 4 + 4 + 36 + 1 + (RegionInfo2_Fields.size * 8)
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleGodUpdateRegionInfo(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 143 (GodUpdateRegionInfo).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x8F).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packVariable(byteBuffer, RegionInfo_Field.SimName, 1)
        packInt(byteBuffer, RegionInfo_Field.EstateID)
        packInt(byteBuffer, RegionInfo_Field.ParentEstateID)
        packInt(byteBuffer, RegionInfo_Field.RegionFlags)
        packFloat(byteBuffer, RegionInfo_Field.BillableFactor)
        packInt(byteBuffer, RegionInfo_Field.PricePerMeter)
        packInt(byteBuffer, RegionInfo_Field.RedirectGridX)
        packInt(byteBuffer, RegionInfo_Field.RedirectGridY)
        byteBuffer.put((RegionInfo2_Fields.size.toByte()))
        for (entry in RegionInfo2_Fields) {
            packLong(byteBuffer, entry.RegionFlagsExtended)
        }
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        RegionInfo_Field.SimName = unpackVariable(byteBuffer, 1)
        RegionInfo_Field.EstateID = unpackInt(byteBuffer)
        RegionInfo_Field.ParentEstateID = unpackInt(byteBuffer)
        RegionInfo_Field.RegionFlags = unpackInt(byteBuffer)
        RegionInfo_Field.BillableFactor = unpackFloat(byteBuffer)
        RegionInfo_Field.PricePerMeter = unpackInt(byteBuffer)
        RegionInfo_Field.RedirectGridX = unpackInt(byteBuffer)
        RegionInfo_Field.RedirectGridY = unpackInt(byteBuffer)
        val i = (byteBuffer.get().toInt() and 0xFF)
        repeat(i) {
            val regionInfo2 = RegionInfo2()
            regionInfo2.RegionFlagsExtended = unpackLong(byteBuffer)
            RegionInfo2_Fields.add(regionInfo2)
        }
    }
}
